### PR TITLE
Handle seenEmittedFiles which was not being set when emit of a file was complete

### LINKED ISSUE #31398: tsc instance reuse in composite project causes out of memory crash
**TypeScript Version:**  3.4.5
**Search Terms:** Composite, project references, out of memory

**Expected behavior:**
Building a large composite project where each component builds successfully should not run out of memory.

**Actual behavior:**
It looks like the same `tsc` instance is being used to compile each phase, and the memory grows until it trips the node memory limit. Would it instead make sense to fork off a new tsc process for each phase, or force some sort of garbage collection?

### PR DESCRIPTION
**Pull request by sheetalkamat**
It was issue only when errors are reported before emitting (which puts the files into pendingEmit that needs to check only in `seenEmittedFiles` and not `seenAffectedFiles`)
If emit happens before semantic diagnostics query this issue is not repro, because the affected files come into play and those are being set correctly.
Fixes #31398

Thanks to @ericanderson and @bjlaub for getting us the repro.

As part of this I also found out that when `--declarationMap` is on in project we were using `.d.ts.map` instead of `.d.ts` as signature which is not desirable (since the map can change without changing d.ts resulting in complete refresh even if it didn't affect .d.ts). I have fixed this in this PR along with ensuring that we have test case for that. Changed the createHash of SolutionBuilderHost for tests to emit `signature-content` so that its easily diagnosable.

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by bjlaub** [comment]
@sheetalkamat I ran this branch on our production codebase test case (from https://github.com/microsoft/TypeScript/pull/33061#issuecomment-524873855) and it seems to fix the issue! Thanks so much for getting this resolved!

**Comment by ericanderson** [comment]
@sheetalkamat Will this be in 3.6.1?

**Comment by sheetalkamat** [comment]
>  Will this be in 3.6.x?

yes

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -214,6 +215,10 @@ namespace ts {
             if (fileExtensionIs(inputFileName, Extension.Dts)) continue;
             const jsFilePath = getOutputJSFileName(inputFileName, configFile, ignoreCase);
             if (jsFilePath) return jsFilePath;
+            if (fileExtensionIs(inputFileName, Extension.Json)) continue;
+            if (getEmitDeclarations(configFile.options) && hasTSFileExtension(inputFileName)) {
```
Comment (sandersn) [inline src/compiler/emitter.ts]: quick note for @weswigham: this new line will probably need when bringing the js-to-dts PR up to date with master

**Comment by sheetalkamat** [comment]
@typescript-bot cherry-pick this to release-3.6

**Pull request merged**